apiserver/objects.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package apiserver
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+type objectsCharmHTTPHandler struct {
+	GetHandler          FailableHandlerFunc
+	LegacyCharmsHandler http.Handler
+}
+
+func (h *objectsCharmHTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var err error
+	switch r.Method {
+	case "GET":
+		err = errors.Annotate(h.GetHandler(w, r), "cannot retrieve charm")
+		if err == nil {
+			// Chain call to legacy (REST API) charms handler
+			h.LegacyCharmsHandler.ServeHTTP(w, r)
+		}
+	default:
+		http.Error(w, fmt.Sprintf("http method %s not implemented", r.Method), http.StatusNotImplemented)
+		return
+	}
+
+	if err != nil {
+		if err := sendJSONError(w, r, errors.Trace(err)); err != nil {
+			logger.Errorf("%v", errors.Annotate(err, "cannot return error to user"))
+		}
+	}
+
+}
+
+// s3CharmsHandler handles charm upload through S3-compatible HTTPS in the API
+// server.
+type objectsCharmHandler struct {
+	ctxt httpContext
+}
+
+// ServeGet serves the GET method for the S3 API. This is the equivalent of the
+// `GetObject` method in the AWS S3 API.
+// Since juju's S3 API only acts as a shim, this method will only rewrite the
+// http request for it to be correctly processed by the legacy /charms handler.
+func (h *objectsCharmHandler) ServeGet(w http.ResponseWriter, r *http.Request) error {
+	logger.Child("s3charmsHandler").Tracef("ServeGet(%s)", r.URL)
```

Comment:
Let's not attempt to create a child logger here for every GET. Instead, cache the child logger in the `objectsCharmHandler`. Also, I don't think it's wise to run this as trace for every request.

```suggestion
	if h.logger.IsTraceEnabled() {
		h.logger.Tracef("ServeGet(%s)", r.URL)
	}
```

Reply:
This was copied from somewhere when I started coding it, but indeed it's not hugely useful, I simply removed this log